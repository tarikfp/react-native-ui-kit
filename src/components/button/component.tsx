import PropTypes from "prop-types";
import * as React from "react";
import {
  ActivityIndicator,
  ActivityIndicatorProps,
  StyleProp,
  StyleSheet,
  Text as RNText,
  TextStyle,
  TouchableOpacity,
  TouchableOpacityProps,
  View,
  ViewStyle,
} from "react-native";
import warnOnce from "warn-once";
import { BaseColors, BaseFonts } from "../../theme";
import { getContrastedTextColor, getHexColor, StylePropType } from "../utils";
import { default as Icon, IconProps } from "../vector-icons";

export type Props = TouchableOpacityProps & {
  /**
   * Overrides the default button style.
   */
  readonly style?: StyleProp<ViewStyle> | undefined;
  /**
   * Size of the button.
   * @default medium
   */
  readonly size?: "large" | "medium" | "small" | "xsmall";
  /**
   * Whether button is in loading state, indicator will be displayed in case of true
   */
  readonly loading?: boolean;
  /**
   * Overrides style for the button text
   */
  readonly labelStyle?: StyleProp<TextStyle>;
  /**
   * Displays text on the button
   */
  readonly label: string;
  /**
   * If provided text, start and end icon colors will be computed based on given backgroundColor.
   * @see Change text color based on brightness of the covered background area
   * https://stackoverflow.com/questions/11867545/change-text-color-based-on-brightness-of-the-covered-background-area
   *
   * Example:
   * Button text, start and end icon (if provided) colors will be `white` in case of the following example below
   *
   * ```js
   *  <Button
   *    backgroundColor="#6666FF"
   *    label="Go back"
   *  />
   * ```
   *
   *
   *
   * It can be overridden by the `style` prop
   *
   * Example: In the following example button background color will be `gold`
   * even though backgroundColor prop is provided.
   *
   *
   * ```js
   *  <Button
   *    size="large"
   *    backgroundColor="#6666FF"
   *    startIcon={{ name: "arrow-left", type: "MaterialCommunityIcons" }}
   *    label="Go back"
   *  />
   * ```
   *
   */
  readonly backgroundColor?: string;
  /**
   * Background color of button when disabled is true
   */
  readonly disabledBackgroundColor?: string;
  /**
   * Left aligned button icon
   * @see IconProps
   */
  readonly startIcon?: IconProps;
  /**
   * Right aligned button icon
   * @see IconProps
   */
  readonly endIcon?: IconProps;
  /**
   * Props of indicator which will be visible on loading state
   * @see ActivityIndicatorProps
   */
  readonly loadingProps?: ActivityIndicatorProps;
};

interface ButtonBaseStyle {
  buttonStyle: StyleProp<ViewStyle>;
  textStyle: StyleProp<TextStyle>;
  iconStyle: { size: number };
}

function Button({
  size = "medium",
  labelStyle,
  label,
  loading,
  disabledBackgroundColor,
  loadingProps,
  backgroundColor,
  startIcon,
  endIcon,
  style,
  ...props
}: Props) {
  // Returns button size, should be one of the following 'large' | 'medium' | 'small'
  // If size does not match, return 'large' size as fallback

  const buttonSizeStyle = (): ButtonBaseStyle => {
    switch (size) {
      case "large":
        return {
          buttonStyle: {
            width: "100%",
            height: 56,
          },
          textStyle: {
            ...BaseFonts.h3,
          },
          iconStyle: {
            size: 25,
          },
        };
      case "medium":
        return {
          buttonStyle: {
            width: "80%",
            height: 48,
          },
          textStyle: {
            ...BaseFonts.h4,
          },
          iconStyle: {
            size: 22,
          },
        };
      case "small":
        return {
          buttonStyle: {
            minWidth: 103,
            height: 40,
            paddingHorizontal: 24,
          },
          textStyle: {
            ...BaseFonts.h4,
          },
          iconStyle: {
            size: 18,
          },
        };
      case "xsmall":
        return {
          buttonStyle: {
            minWidth: 83,
            height: 35,
            paddingHorizontal: 16,
          },
          textStyle: {
            ...BaseFonts.p1,
          },
          iconStyle: {
            size: 15,
          },
        };
      default:
        warnOnce(
          true,
          `Invalid button size. Should be one of the following... "large" | "medium" | "small" | "xsmall"`
        );

        return {
          buttonStyle: {
            minWidth: 103,
            height: 40,
            paddingHorizontal: 24,
          },
          textStyle: {
            ...BaseFonts.h4,
          },
          iconStyle: {
            size: 18,
          },
        };
    }
  };

  // prevent recomputing contrast color on every render
  // in order to avoid flickering on text color
  const textContrastColor = React.useMemo(() => {
    if (backgroundColor) {
      return getContrastedTextColor(getHexColor(backgroundColor));
    }
  }, [backgroundColor]);

  return (
    <TouchableOpacity
      activeOpacity={props.activeOpacity ?? 0.5}
      style={StyleSheet.flatten<ViewStyle>([
        buttonSizeStyle().buttonStyle,
        styles.btnBase,
        {
          backgroundColor: props.disabled
            ? disabledBackgroundColor ?? backgroundColor
            : backgroundColor,
        },
        style,
      ])}
      {...props}>
      <View style={styles.innerContainer}>
        {!loading && startIcon && (
          <Icon
            style={StyleSheet.flatten<ViewStyle>([
              { marginRight: 6 },
              startIcon.style,
            ])}
            size={buttonSizeStyle().iconStyle.size}
            color={textContrastColor}
            {...startIcon}
          />
        )}

        <RNText
          style={StyleSheet.flatten<TextStyle>([
            { color: textContrastColor },
            buttonSizeStyle().textStyle,
            labelStyle,
          ])}>
          {label}
        </RNText>

        {loading ? (
          <ActivityIndicator
            style={StyleSheet.flatten<ViewStyle>([
              styles.indicator,
              loadingProps?.style,
            ])}
            size={buttonSizeStyle().iconStyle.size}
            color={textContrastColor}
            {...loadingProps}
          />
        ) : (
          endIcon && (
            <Icon
              style={StyleSheet.flatten<TextStyle>([
                { marginLeft: 6 },
                endIcon.style,
              ])}
              size={buttonSizeStyle().iconStyle.size}
              color={textContrastColor}
              {...endIcon}
            />
          )
        )}
      </View>
    </TouchableOpacity>
  );
}

export default Button;

Button.propTypes = {
  style: StylePropType,
  size: PropTypes.oneOf(["large", "medium", "small", "xsmall"]),
  loading: PropTypes.bool,
  labelStyle: StylePropType,
  label: PropTypes.string,
  backgroundColor: PropTypes.string,
  disabledBackgroundColor: PropTypes.string,
  startIcon: PropTypes.object,
  endIcon: PropTypes.object,
  indicatorProps: PropTypes.object,
};

interface Styles {
  btnBase: ViewStyle;
  indicator: ViewStyle;
  innerContainer: ViewStyle;
}

const styles = StyleSheet.create<Styles>({
  btnBase: {
    borderRadius: 6,
    justifyContent: "center",
    padding: 8,
    alignItems: "center",
    borderWidth: 1,
    borderColor: BaseColors.borderColor,
    backgroundColor: BaseColors.primary,
  },
  indicator: {
    marginLeft: 14,
  },
  innerContainer: {
    flexDirection: "row",
    alignItems: "center",
    justifyContent: "center",
  },
});
