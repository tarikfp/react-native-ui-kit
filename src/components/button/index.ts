export { default as Button, Props as ButtonProps } from "./component";
