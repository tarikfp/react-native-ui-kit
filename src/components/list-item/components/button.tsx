import PropTypes from "prop-types";
import * as React from "react";
import { StyleProp, View, ViewStyle } from "react-native";
import { Button, ButtonProps } from "../../button";
import { StylePropType } from "../../utils";

export type Props = ButtonProps & {
  /**
   * style of the button wrapper
   */
  wrapperStyle?: StyleProp<ViewStyle>;
  /**
   * alignment of the button
   */
  alignment?: "start" | "end";
};

export default function ListItemButton({
  alignment,
  wrapperStyle,
  ...buttonProps
}: Props) {
  return (
    <View style={[getWrapperStyle(alignment), wrapperStyle]}>
      <Button size="xsmall" {...buttonProps} />
    </View>
  );
}

const getWrapperStyle = (
  alignment?: "start" | "end"
): StyleProp<ViewStyle> => ({
  flex: alignment ? 1 : 0,
  flexDirection: "row",
  justifyContent:
    alignment === "start"
      ? "flex-start"
      : alignment === "end"
      ? "flex-end"
      : "center",
});

ListItemButton.displayName = "ListItem.Button";

ListItemButton.propTypes = {
  style: StylePropType,
  alignment: PropTypes.oneOf(["start", "end"]),
};
