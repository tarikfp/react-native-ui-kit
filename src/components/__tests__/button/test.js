import { fireEvent, render } from "@testing-library/react-native";
import * as React from "react";
import renderer from "react-test-renderer";
import { Button } from "../../button";

describe("tests button component", () => {
  it("renders button with label", () => {
    const tree = renderer.create(<Button label="button label" />).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it("renders button with size", () => {
    const tree = renderer
      .create(<Button label="Test" size="medium" />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it("renders button with loading state", () => {
    const tree = renderer
      .create(<Button loading loadingProps={{ testID: "button-loading" }} />)
      .toJSON();

    expect(tree).toMatchSnapshot();

    const { getByTestId } = render(
      <Button loading loadingProps={{ testID: "button-loading" }} />
    );

    expect(getByTestId("button-loading")).toBeDefined();
  });

  it("renders button with custom label style", () => {
    const tree = renderer
      .create(
        <Button label="Test" labelStyle={{ fontSize: 24, fontWeight: "700" }} />
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it("renders button with custom background color", () => {
    const tree = renderer
      .create(<Button label="Test" backgroundColor="orange" />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it("renders button with custom disabled background color", () => {
    const tree = renderer
      .create(
        <Button label="Test" disabled disabledBackgroundColor="slategray" />
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it("renders button with custom start icon", () => {
    const tree = renderer
      .create(
        <Button
          label="Test"
          startIcon={{ name: "arrow-left", type: "MaterialCommunityIcons" }}
        />
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it("renders button with custom end icon", () => {
    const tree = renderer
      .create(
        <Button
          label="Test"
          endIcon={{ name: "arrow-right", type: "MaterialCommunityIcons" }}
        />
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it("should execute onPress", () => {
    const onPressMock = jest.fn();

    const { getByTestId } = render(
      <Button
        testID="button"
        onPress={onPressMock}
        loading
        loadingProps={{ testID: "button-loading" }}
      />
    );

    fireEvent(getByTestId("button"), "onPress");

    expect(onPressMock).toHaveBeenCalledTimes(1);
  });

  it("should execute onPressIn", () => {
    const onPressInMock = jest.fn();

    const { getByTestId } = render(
      <Button
        testID="button"
        onPressIn={onPressInMock}
        loading
        loadingProps={{ testID: "button-loading" }}
      />
    );

    fireEvent(getByTestId("button"), "onPressIn");

    expect(onPressInMock).toHaveBeenCalledTimes(1);
  });

  it("should execute onPressOut", () => {
    const onPressOutMock = jest.fn();

    const { getByTestId } = render(
      <Button
        testID="button"
        onPressOut={onPressOutMock}
        loading
        loadingProps={{ testID: "button-loading" }}
      />
    );

    fireEvent(getByTestId("button"), "onPressOut");

    expect(onPressOutMock).toHaveBeenCalledTimes(1);
  });
});
