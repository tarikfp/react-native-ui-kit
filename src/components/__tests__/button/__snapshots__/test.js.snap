// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests button component renders button with custom background color 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "orange",
      "borderColor": "rgba(0, 0, 0,0.05)",
      "borderRadius": 6,
      "borderWidth": 1,
      "height": 48,
      "justifyContent": "center",
      "opacity": 1,
      "padding": 8,
      "width": "80%",
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "center",
      }
    }
  >
    <Text
      style={
        Object {
          "color": "white",
          "fontSize": 14,
          "fontWeight": "500",
          "lineHeight": 20,
        }
      }
    >
      Test
    </Text>
  </View>
</View>
`;

exports[`tests button component renders button with custom disabled background color 1`] = `
<View
  accessibilityState={
    Object {
      "disabled": true,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "slategray",
      "borderColor": "rgba(0, 0, 0,0.05)",
      "borderRadius": 6,
      "borderWidth": 1,
      "height": 48,
      "justifyContent": "center",
      "opacity": 1,
      "padding": 8,
      "width": "80%",
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "center",
      }
    }
  >
    <Text
      style={
        Object {
          "color": undefined,
          "fontSize": 14,
          "fontWeight": "500",
          "lineHeight": 20,
        }
      }
    >
      Test
    </Text>
  </View>
</View>
`;

exports[`tests button component renders button with custom end icon 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": undefined,
      "borderColor": "rgba(0, 0, 0,0.05)",
      "borderRadius": 6,
      "borderWidth": 1,
      "height": 48,
      "justifyContent": "center",
      "opacity": 1,
      "padding": 8,
      "width": "80%",
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "center",
      }
    }
  >
    <Text
      style={
        Object {
          "color": undefined,
          "fontSize": 14,
          "fontWeight": "500",
          "lineHeight": 20,
        }
      }
    >
      Test
    </Text>
    <Text
      allowFontScaling={false}
      selectable={false}
      style={
        Array [
          Object {
            "color": "#757575",
            "fontSize": 22,
          },
          Object {
            "marginLeft": 6,
          },
          Object {
            "fontFamily": "Material Design Icons",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          Object {},
        ]
      }
    >
      󰁔
    </Text>
  </View>
</View>
`;

exports[`tests button component renders button with custom label style 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": undefined,
      "borderColor": "rgba(0, 0, 0,0.05)",
      "borderRadius": 6,
      "borderWidth": 1,
      "height": 48,
      "justifyContent": "center",
      "opacity": 1,
      "padding": 8,
      "width": "80%",
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "center",
      }
    }
  >
    <Text
      style={
        Object {
          "color": undefined,
          "fontSize": 24,
          "fontWeight": "700",
          "lineHeight": 20,
        }
      }
    >
      Test
    </Text>
  </View>
</View>
`;

exports[`tests button component renders button with custom start icon 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": undefined,
      "borderColor": "rgba(0, 0, 0,0.05)",
      "borderRadius": 6,
      "borderWidth": 1,
      "height": 48,
      "justifyContent": "center",
      "opacity": 1,
      "padding": 8,
      "width": "80%",
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "center",
      }
    }
  >
    <Text
      allowFontScaling={false}
      selectable={false}
      style={
        Array [
          Object {
            "color": "#757575",
            "fontSize": 22,
          },
          Object {
            "marginRight": 6,
          },
          Object {
            "fontFamily": "Material Design Icons",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          Object {},
        ]
      }
    >
      󰁍
    </Text>
    <Text
      style={
        Object {
          "color": undefined,
          "fontSize": 14,
          "fontWeight": "500",
          "lineHeight": 20,
        }
      }
    >
      Test
    </Text>
  </View>
</View>
`;

exports[`tests button component renders button with label 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": undefined,
      "borderColor": "rgba(0, 0, 0,0.05)",
      "borderRadius": 6,
      "borderWidth": 1,
      "height": 48,
      "justifyContent": "center",
      "opacity": 1,
      "padding": 8,
      "width": "80%",
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "center",
      }
    }
  >
    <Text
      style={
        Object {
          "color": undefined,
          "fontSize": 14,
          "fontWeight": "500",
          "lineHeight": 20,
        }
      }
    >
      button label
    </Text>
  </View>
</View>
`;

exports[`tests button component renders button with loading state 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": undefined,
      "borderColor": "rgba(0, 0, 0,0.05)",
      "borderRadius": 6,
      "borderWidth": 1,
      "height": 48,
      "justifyContent": "center",
      "opacity": 1,
      "padding": 8,
      "width": "80%",
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "center",
      }
    }
  >
    <Text
      style={
        Object {
          "color": undefined,
          "fontSize": 14,
          "fontWeight": "500",
          "lineHeight": 20,
        }
      }
    />
    <ActivityIndicator
      size={22}
      style={
        Object {
          "marginLeft": 14,
        }
      }
      testID="button-loading"
    />
  </View>
</View>
`;

exports[`tests button component renders button with size 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": undefined,
      "borderColor": "rgba(0, 0, 0,0.05)",
      "borderRadius": 6,
      "borderWidth": 1,
      "height": 48,
      "justifyContent": "center",
      "opacity": 1,
      "padding": 8,
      "width": "80%",
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "center",
      }
    }
  >
    <Text
      style={
        Object {
          "color": undefined,
          "fontSize": 14,
          "fontWeight": "500",
          "lineHeight": 20,
        }
      }
    >
      Test
    </Text>
  </View>
</View>
`;
